lib/JSON/Validator.pm

```diff
@@ -573,69 +573,85 @@ sub _validate {
 
 sub _validate_all_of {
   my ($self, $data, $path, $rules) = @_;
-  my $type = data_type $data, $rules;
-  my (@errors, @expected);
+  my (@errors);
 
   my $i = 0;
   for my $rule (@$rules) {
     next unless my @e = $self->_validate($_[1], $path, $rule);
-    my $schema_type = schema_type $rule;
-    push @expected, $schema_type if $schema_type;
-    push @errors, [$i, @e] if !$schema_type or $schema_type eq $type;
+    push @errors, [$i, @e];
   }
   continue {
     $i++;
   }
 
-  return E $path, [allOf => type => join('/', uniq @expected), $type]
-    if !@errors and @expected;
-  return prefix_errors allOf => @errors if @errors;
-  return;
+  return if not @errors;
+
+  return E $path,
+    [
+    allOf => type => join('/', uniq map $_->[1]->details->[0], @errors),
+    $errors[-1][1]->details->[2]
+    ]
+    if @errors > 1 and not grep $_->[1]->details->[1] ne 'type', @errors;
+
+  return prefix_errors allOf => @errors;
 }
 
 sub _validate_any_of {
-  my ($self, $data, $path, $rules) = @_;
-  my $type = data_type $data, $rules;
-  my (@e, @errors, @expected);
+  my ($self, $data, $path, $rules, $prefix) = @_;
+  $prefix //= 1;
+  my (@e, @errors);
 
   my $i = 0;
   for my $rule (@$rules) {
     @e = $self->_validate($_[1], $path, $rule);
+
+    # TODO: the spec might not think that short-circuiting is ok here
     return unless @e;
-    my $schema_type = schema_type $rule;
-    push @errors, [$i, @e] and next if !$schema_type or $schema_type eq $type;
-    push @expected, $schema_type;
+    push @errors, [$i, @e];
   }
   continue {
     $i++;
   }
 
-  my $expected = join '/', uniq @expected;
-  return E $path, [anyOf => type => $expected, $type] unless @errors;
-  return prefix_errors anyOf => @errors;
+  if (grep $_->[1]->details->[1] ne 'type', @errors) {
+    return $prefix ? prefix_errors(anyOf => @errors) : (map $_->[1], @errors);
+  }
+
+  my $want_types = join('/', uniq map $_->[1]->details->[0], @errors);
+  return E $path,
+    [
+    ($prefix ? (anyOf => type => $want_types) : ($want_types => 'type')),
+    $errors[-1][1]->details->[2]
+    ];
```

Comment:
Pretty sure this can be made short as well. See my comment above about "This is very difficult to read". (I'm thinking "extract `@details` first) This goes for `_validate_one_of()` below as well.